src/libmongoc/tests/test-awsauth.c

```diff
@@ -0,0 +1,515 @@
+                               __FILE__,                                       \
+                               __LINE__,                                       \
+                               BSON_FUNC,                                      \
+                               BSON_STR (Cond));                               \
+         MONGOC_STDERR_PRINTF ("MESSAGE: " Fmt "\n", __VA_ARGS__);             \
+         abort ();                                                             \
+      }                                                                        \
+   } else                                                                      \
+      ((void) 0)
+
+#define FAILF(Fmt, ...)                                                     \
+   if (1) {                                                                 \
+      MONGOC_STDERR_PRINTF ("FAIL:%s:%d  %s()\n  Condition '%s' failed.\n", \
+                            __FILE__,                                       \
+                            __LINE__,                                       \
+                            BSON_FUNC,                                      \
+                            BSON_STR (Cond));                               \
+      MONGOC_STDERR_PRINTF ("MESSAGE: " Fmt "\n", __VA_ARGS__);             \
+      abort ();                                                             \
+   } else                                                                   \
+      ((void) 0)
+
+static void
+test_auth (mongoc_database_t *db, bool expect_failure)
+{
+   bson_error_t error;
+   bson_t *ping = BCON_NEW ("ping", BCON_INT32 (1));
+   bool ok = mongoc_database_command_with_opts (db,
+                                                ping,
+                                                NULL /* read_prefs */,
+                                                NULL /* opts */,
+                                                NULL /* reply */,
+                                                &error);
+   if (expect_failure) {
+      ASSERTF (!ok, "%s", "Expected auth failure, but got success");
+   } else {
+      ASSERTF (ok, "Expected auth success, but got error: %s", error.message);
+   }
+   bson_destroy (ping);
+}
+
+// creds_eq returns true if `a` and `b` contain the same credentials.
+static bool
+creds_eq (_mongoc_aws_credentials_t *a, _mongoc_aws_credentials_t *b)
+{
+   BSON_ASSERT_PARAM (a);
+   BSON_ASSERT_PARAM (b);
+
+   if (0 != strcmp (a->access_key_id, b->access_key_id)) {
+      return false;
+   }
+   if (0 != strcmp (a->secret_access_key, b->secret_access_key)) {
+      return false;
+   }
+   if (0 != strcmp (a->session_token, b->session_token)) {
+      return false;
+   }
+   if (a->expiration.set != b->expiration.set) {
+      return false;
+   }
+   if (a->expiration.set) {
+      if (mcd_time_compare (a->expiration.value.expire_at,
+                            b->expiration.value.expire_at) != 0) {
+         return false;
+      }
+   }
```

Comment:
Confirming, is the expiration date supposed to be considered part of credentials equality?